Create rule S6563: Use UTC when recoding DateTime instants. (#1713)
SonarSource/rspec: rules/S6563/csharp/rule.adoc

== Why is this an issue?

> The property `DateTime.Now` returns the current date and time expressed in the machine's local time without containing any timezone-related data (e.g., offset).

Reviewer: picky thing, but we avoid Latin abbreviations in tech documentation, e.g. can be read as egg by screen readers. I'd change this for example offsets

> The property `DateTime.Now` returns the current date and time expressed in the machine's local time without containing any timezone-related data (for example the offset).

Reviewer 2: It's only for `DateTime.Now`.
`DateTimeOffset.Now` contains the offset so it's clear at that instant how far away you were from UTC.

> Not having this data means that if you need to display this `DateTime` object or use it for computations in another machine placed in a different time zone, you won't be able to reconstruct it in the second machine's local time without knowing the origin's offset. This will likely lead to confusion and potential bugs.

Reviewer 3: What do you think about using information instead of data in this and the previous sentence?
Data is something raw that does not carry meaningful information for the user, in this example I feel we should call it information, as it is important for the user to properly process the DateTime object.

Reviewer 2: Sounds good! I'll change it.

Instead, you should record the `DateTime` instants in UTC, which gives you the date and time as it is in the Coordinated Universal Time. Daylight Saving Time (DST) is not used for UTC as it is a time standard for all time zones.

Generally, unless you want to only display the Date and Time to a user on their local machine, you should always use UTC (foe example when storing dates in a DB, using them for calculations).

=== What is the potential impact?

> You can end up with `DateTime` instants that have no meaning for anyone except the machine they were recorded in.

Reviewer 3: ```suggestion
You can end up with `DateTime` instants that have no meaning for anyone except the machine they were recorded on.
```

> Additionally, by using UTC, you get an unambiguous representation of an instant, as you can get from any UTC instant to the equivalent local time. This operation isn't reversible as some local times are ambiguous and can be matched to more than one UTC instant (for example, due to daylight savings).

Reviewer: Take out additionally, puts a greater emphasis on your point "By using UTC"

What do you mean by "as you can get from any UTC instant to the equivalent local time." This second half of the sentence is a tad confusing

Reviewer 2: Thanks, Cyn, I rephrased hoping it's clearer.

What I mean is that if I record a DateTime in UTC - for example `01/03/2023 00:00 +0` I can transform it to any "local" time I want.  For example, this in Geneva time will be `01/03/2023 02:00 +2` .

> By using UTC, you get an unambiguous representation of an instant, as you can get from any UTC instant to the equivalent local time. This operation isn't reversible as some local times are ambiguous and can be matched to more than one UTC instant (for example, due to daylight savings).

Reviewer 3: No strong opinion this one, though I will let it here and let you decide if you like it more or not:

Using UTC gives an unambiguous representation of an instant, and this UTC instant can be transformed into any equivalent local time.

Reviewer 2: Simpler and better.
Thanks!

== How to fix it

The fix is to use `DateTime.UtcNow`, `DateTimeOffSet.Now` (as it contains offset information) or, `DateTimeOffSet.UtcNow`.

=== Code examples

==== Noncompliant code example

> [source,csharp,diff-id=3,diff-type=noncompliant]

Reviewer 3: Education: Is the diff-id the line number that needs to be marked as a diff? And what if there are multiple lines?

Reviewer 2: No idea - we need more info on how this works :/ 
Also, I don't know how to visualize this to know if it's correct or wrong.

Reviewer 3: Ok, I think actually what we need is this documentation https://github.com/SonarSource/rspec/blob/master/docs/description.adoc#diff-view.
If I understand it correctly, the ID is a unique identifier to tell the products which two snippets to compare, and they will find the diff and colour accordingly. 
So, in your case the ID can be 1

----
void LogDateTime()
{
    using var streamWriter = new StreamWriter("logs.txt", true);

>     streamWriter.WriteLine($"DateTime:{DateTimeNow.ToString("o");}"); // This log won't have any meaning if it's reconstructed in a machine in a different timezone.

Reviewer 3: ```suggestion
    streamWriter.WriteLine($"DateTime:{DateTime.Now.ToString("o");}"); // This log won't have any meaning if it's reconstructed in a machine in a different timezone.
```

== Why is this an issue?

The property `DateTime.Now` returns the current date and time expressed in the machine's local time without containing any timezone-related data (e.g., offset).
Not having this data means you cannot reconstruct the time in another time zone and potentially display it or use it in a machine other than the one in which it was recorded.

> Even if possible, it wouldn't make a difference since `DateTime.Kind` stops being `Local` and turns to `Unspecified` as soon as it crosses a machine's boundary.

Reviewer 3: If I understand correctly, it does not have to turn unspecified all the time. Take a look at this page:
https://learn.microsoft.com/en-us/dotnet/standard/base-types/how-to-round-trip-date-and-time-values

Reviewer 2: I rewrote this - to make it clear that you can reconstruct it - but not in another machine's local time in case its placed in another timezone.

> Instead, you should record the `DateTime` instants in UTC, which tells you the date and time as it would be in Coordinated Universal Time that is not affected by changes such as daylight savings.

Reviewer 3: ```suggestion
Instead, you should record the `DateTime` instants in UTC, which gives you the date and time as it is in the Coordinated Universal Time. Because it always returns the time from the same timezone, it is not affected by changes such as daylight savings.
```

Reviewer 2: UTC is a time standard and not a timezone - so IMO I should not apply the last sentence change unless I've understood totally wrong.

Generally, unless you want to only display the Date and Time to a user on their local machine, you should always use UTC (e.g., when storing dates in a DB, using them for calculations).

=== What is the potential impact?

You can end up with `DateTime` instants that have no meaning for anyone except the machine they were recorded in.
Additionally, by using UTC, you get an unambiguous representation of an instant, as you can get from any UTC instant to the equivalent local time. This operation isn't reversible as some local times are ambiguous and can be matched to more than one UTC instant (for example, due to daylight savings).

== How to fix it

The fix is to use `DateTime.UtcNow`, `DateTimeOffSet.Now` (as it contains offset information) or, `DateTimeOffSet.UtcNow`.

=== Code examples

==== Noncompliant code example

[source,csharp,diff-id=3,diff-type=noncompliant]
----
void LogDateTime()
{
    using var streamWriter = new StreamWriter("logs.txt", true);

>     streamWriter.WriteLine($"DateTime:{DateTime.Now.ToString("dd MMM yyyy HH: mm:ss")}"); // This log won't have any meaning outside the machine it was recorded in

Reviewer 3: As we plan to add a rule that says don`t hardcode the format of the date time, I would not use it in these examples either.

Reviewer 2: I followed the example from the docs more or less to make it more realistic.
